### PR TITLE
[SPARK-39728][PYTHON] Add explicit PySpark SQL function parity check

### PR DESCRIPTION
**Pull request by aray**
### What changes were proposed in this pull request?

This PR adds a test that compares the available list of Python DataFrame functions in pyspark.sql.functions with those available in the Scala/Java DataFrame API in org.apache.spark.sql.functions. If a function is added to only one but not the other this test will fail until its exclusions are updated.

### Why are the changes needed?

Currently there is no easy way to verify what functions are missing from the Python DataFrame API

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

This PR is testing only

### CODE REVIEW

**Comment by huaxingao** [comment]
cc @HyukjinKwon

File: python/pyspark/sql/tests/test_functions.py [hunk 1]
```diff
@@ -55,6 +55,74 @@
...
+        #     pyspark.sql.functions or added to the below expected_missing_in_py set.
+        # * If a function was added to pyspark.sql.functions that was already in
+        #     org.apache.spark.sql.functions then it needs to be removed from expected_missing_in_py
+        #     below. If the function has a different name it needs to be added to py_equiv_jvm
+        #     mapping.
+        # * If it's not related to an added/removed function then likely one of the exclusion lists
+        #     need updated.
+        from pyspark import SparkContext
+        from pyspark.sql import functions as py_functions
+        from inspect import getmembers, isfunction
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_functions.py:74]: Let's just import it on the top

File: python/pyspark/sql/tests/test_functions.py [hunk 2]
```diff
@@ -55,6 +55,74 @@
...
+            "callUDF",  # depreciated, use call_udf
+            "typedlit",  # Scala only
+            "typedLit",  # Scala only
+            "monotonicallyIncreasingId",  # depreciated, use monotonically_increasing_id
+            "negate",  # equivalent to python -expression
+            "not",  # equivalent to python ~expression
+        ]
+        # Excluded functions on the python side
+        py_excluded_fn = [
+            "pandas_udf",  # Python only
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_functions.py:96]: I think we don't need to test this case. It's pretty unlikely there are some PySpark API that are not implemented mistakenly only in JVM

File: python/pyspark/sql/tests/test_functions.py (hunk 1 above)
Comment (aray) [reply python/pyspark/sql/tests/test_functions.py]: I moved the imports to the top of the function. Not sure if you wanted that or global imports. I do see plenty of other tests that have some test specific imports at the beginning of their function.

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for making a PR, @aray .

File: python/pyspark/sql/tests/test_functions.py (hunk 1 above)
Comment (HyukjinKwon) [reply python/pyspark/sql/tests/test_functions.py]: yeah, I know what you're doing is consistent with some places but I am trying to just move all imports as a global. I think we should just put all imports together as global ones.

Comment (aray) [reply python/pyspark/sql/tests/test_functions.py]: no problem, done

File: python/pyspark/sql/tests/test_functions.py [hunk 3]
```diff
@@ -55,6 +55,62 @@
...
+        py_equiv_jvm = {"create_map": "map"}
+        for py_name, jvm_name in py_equiv_jvm.items():
+            if py_name in py_fn_set:
+                py_fn_set.remove(py_name)
+                py_fn_set.add(jvm_name)
+
+        missing_in_py = jvm_fn_set.difference(py_fn_set)
+
+        # Functions that we expect to be missing in python until they are added to pyspark
+        # TODO Add all of these functions to pyspark
```
[the remark this answers is not in the document]
Comment (aray) [reply python/pyspark/sql/tests/test_functions.py]: I've gone through the list and found existing or created new JIRA's for each and updated TODO's as such.

**Review by HyukjinKwon — APPROVED** [review]

File: python/pyspark/sql/tests/test_functions.py [hunk 4]
```diff
@@ -51,10 +52,65 @@
...
+                py_fn_set.add(jvm_name)
+
+        missing_in_py = jvm_fn_set.difference(py_fn_set)
+
+        # Functions that we expect to be missing in python until they are added to pyspark
+        expected_missing_in_py = {
+            "call_udf",  # TODO(SPARK-39734)
+            "localtimestamp",  # TODO(SPARK-36259)
+            "map_contains_key",  # TODO(SPARK-39733)
+            "pmod",  # TODO(SPARK-37348)
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/test_functions.py:107]: cc @itholic @xinrong-databricks @yoda-mon in case you're interested in them.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Review by dongjoon-hyun — COMMENTED** [review]
+1, LGTM, too. Thank you, @aray , @HyukjinKwon , @huaxingao .

**Pull request closed without merging**